docs/aspnetcore-facility.md

```diff
@@ -10,38 +10,78 @@ The ASP.NET Core facility provides Castle Windsor integration using a custom act
-If your controllers, view components or taghelpers are in a different assembly, you call this extension multiple times supplying an example
-type for which Windsor will then scan that assembly for framework types.
+### Registering Controllers, ViewComponents and TagHelpers
 
-If you would like to change the lifestyles of any of these components, then you do not call the `app.UseCastleWindsor<Startup>(container)`
-extension. You can opt for your own conventional registrations instead like so:
+`Controllers`, `ViewComponents` and `TagHelpers` are registered automatically for you when the `services.AddCastleWindsor(container)` extension
+is called. All components are registered with a scoped lifestyle, and are assumed to be a single instance for the duration of the web request. 
+You can optionally override the lifestyles for each of these if you prefer using the `options` callback in the example below:
 
 ```csharp
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<Controller>().LifestyleScoped()); // <- Change lifestyle
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<ViewComponent>().LifestyleTransient());
-container.Register(Classes.FromAssemblyInThisApplication(typeof(TStartup).Assembly)
-	.BasedOn<TagHelper>().LifestyleTransient());
+services.AddCastleWindsor(container, opts =>
+{
+	opts.RegisterControllers(typeof(HomeController).Assembly, LifestyleType.Transient);
+	opts.RegisterTagHelpers(typeof(EmailTagHelper).Assembly, LifestyleType.Transient);
+	opts.RegisterViewComponents(typeof(AddressComponent).Assembly, LifestyleType.Transient);
+});
+```
+
+This is also useful if your framework components live in a separate assemblies or are not defined in the same assembly as your web application.
+Alternatively if your framework components all live in one assembly and you dont need to change lifestyles then you can simply use the following:
+
+```csharp
+services.AddCastleWindsor(container, opts =>
+{
+	opts.UseEntryAssembly(typeof(HomeController).Assembly);
+});
+```
+This is good for trouble shooting situations where nothing get's registered because of problems in the hosting environment where 
+GetEntryAssembly/GetCallingAssembly does not work as expected. 
+
+### Registering custom middleware
+
+If you don't have any DI requirements in your middleware you can simply register it in the IServiceCollection like so: 
+
+```csharp
+services.AddSingleton<FrameworkMiddleware>(); // Do this if you don't care about using Windsor
+```
+
+Alternatively if you would like to inject dependencies into your middleware that are registered with Castle Windsor then you can use the 
+`UseCastleWindsorMiddleware` extension.
+
+```csharp
+// Add custom middleware, do this if your middleware uses DI from Windsor
+app.UseCastleWindsorMiddleware<CustomMiddleware>(container);
```

Comment:
`UseCastleWindsorMiddleware` sounds like it is adding some special Windsor middleware to the pipeline, I was wondering if it should be named something like `UseMiddlewareViaWindsor<TMiddleware>` or `UseMiddlewareFromWindsor<TMiddleware>`?

I know this was there from the first cut but now is the best time to make the change. I was wondering what Autofac named theirs and I can't find how they do this same sort of thing. @fir3pho3nixx you are the expert here so I'll let you decide whether you want to rename it.

Reply:
@jonorossi I really like this. It makes it discoverable via code completion. 

I love `UseMiddlewareViaWindsor<TMiddleware>`, the responsible choice is definitely `UseMiddlewareFromWindsor<TMiddleware>`.

I am happy with how we are refining the UX here.

Reply:
Great, especially that it'll be more discoverable.

Were you planning to name it `UseMiddlewareFromWindsor` or `UseMiddlewareFromCastleWindsor`. I was wondering if the "Castle" bit should be dropped from `AddCastleWindsor` and `UseCastleWindsor` since it doesn't really add anything useful?

Reply:
You will be pleased to know, `UseCastleWindsor` is not a thing anymore. We now only need `UseWindsor` :). Done.

Reply:
Oh and `UseMiddlewareFromWindsor` is the new name for the middleware extension.